lib/location/method_gnss.c

```diff
@@ -0,0 +1,790 @@
+		location_result.longitude = pvt_data.longitude;
+		location_result.accuracy = pvt_data.accuracy;
+		location_result.datetime.valid = true;
+		location_result.datetime.year = pvt_data.datetime.year;
+		location_result.datetime.month = pvt_data.datetime.month;
+		location_result.datetime.day = pvt_data.datetime.day;
+		location_result.datetime.hour = pvt_data.datetime.hour;
+		location_result.datetime.minute = pvt_data.datetime.minute;
+		location_result.datetime.second = pvt_data.datetime.seconds;
+		location_result.datetime.ms = pvt_data.datetime.ms;
+	}
+
+	if (fixes_remaining <= 0) {
+		/* We are done, stop GNSS and publish the fix. */
+		method_gnss_cancel();
+		location_core_event_cb(&location_result);
+	}
+}
+
+#if defined(CONFIG_LOCATION_METHOD_GNSS_AGPS_EXTERNAL)
+static void method_gnss_agps_ext_work_fn(struct k_work *item)
+{
+	location_core_event_cb_assistance_request(&agps_request);
+}
+#endif
+
+static void method_gnss_positioning_work_fn(struct k_work *work)
+{
+	int err = 0;
+	struct method_gnss_start_work_args *work_data =
+		CONTAINER_OF(work, struct method_gnss_start_work_args, work_item);
+	const struct location_gnss_config gnss_config = work_data->gnss_config;
+
+	if (!method_gnss_allowed_to_start()) {
+		/* Location request was cancelled while waiting for RRC idle or PSM. Do nothing. */
+		return;
+	}
+
+	/* Configure GNSS to continuous tracking mode */
+	err = nrf_modem_gnss_fix_interval_set(1);
+
+	/* By default we take the first fix. */
+	fixes_remaining = 1;
+
+	uint8_t use_case;
+
+	switch (gnss_config.accuracy) {
+	case LOCATION_ACCURACY_LOW:
+		use_case = NRF_MODEM_GNSS_USE_CASE_MULTIPLE_HOT_START |
+			   NRF_MODEM_GNSS_USE_CASE_LOW_ACCURACY;
+		err |= nrf_modem_gnss_use_case_set(use_case);
+		break;
+
+	case LOCATION_ACCURACY_NORMAL:
+		use_case = NRF_MODEM_GNSS_USE_CASE_MULTIPLE_HOT_START;
+		err |= nrf_modem_gnss_use_case_set(use_case);
+		break;
+
+	case LOCATION_ACCURACY_HIGH:
+		use_case = NRF_MODEM_GNSS_USE_CASE_MULTIPLE_HOT_START;
+		err |= nrf_modem_gnss_use_case_set(use_case);
+
+		/* In high accuracy mode, use the configured fix count. */
+		fixes_remaining = gnss_config.num_consecutive_fixes;
+		break;
+	}
+
+	if (err) {
+		LOG_ERR("Failed to configure GNSS");
+		location_core_event_cb_error();
+		running = false;
+		return;
+	}
+
+	err = nrf_modem_gnss_start();
+	if (err) {
+		LOG_ERR("Failed to start GNSS");
+		location_core_event_cb_error();
+		running = false;
+		return;
+	}
+
+	location_core_timer_start(gnss_config.timeout);
+}
+
+int method_gnss_location_get(const struct location_method_config *config)
+{
+	const struct location_gnss_config gnss_config = config->gnss;
+	int err;
+
+	/* GNSS event handler is already set once in method_gnss_init(). If no other thread is
+	 * using GNSS, setting it again is not needed.
+	 */
+	err = nrf_modem_gnss_event_handler_set(method_gnss_event_handler);
+	if (err) {
+		LOG_ERR("Failed to set GNSS event handler, error %d", err);
+		return -err;
```

Comment:
From `libmodem v1.4.0` this API (all `nrf_modem_gnss_` APIs) will return a negative `nrf_errno` directly.

Reply:
Also, the values of `nrf_errno` will be the same as the newlibc `errno` values.

Reply:
I'll remove the minus sign.